// replicate-stainless/src/resources/models/readme.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../../core/resource';
import { APIPromise } from '../../core/api-promise';
import { buildHeaders } from '../../internal/headers';
import { RequestOptions } from '../../internal/request-options';
import { path } from '../../internal/utils/path';

export class Readme extends APIResource {
  /**
   * Get the README content for a model.
   *
   * Example cURL request:
   *
   * ```console
   * curl -s \
   *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
   *   https://api.replicate.com/v1/models/replicate/hello-world/readme
   * ```
   *
   * The response will be the README content as plain text in Markdown format:
   *
   * ```
   * # Hello World Model
   *
   * This is an example model that...
   * ```
   *
   * @example
   * ```ts
   * const readme = await replicate.models.readme.get({
   *   model_owner: 'model_owner',
   *   model_name: 'model_name',
   * });
   * ```
   */
  get(params: ReadmeGetParams, options?: RequestOptions): APIPromise<string> {
    const { model_owner, model_name } = params;
    return this._client.get(path`/models/${model_owner}/${model_name}/readme`, {
      ...options,
      headers: buildHeaders([{ Accept: 'text/plain' }, options?.headers]),
    });
  }
}

/**
 * The README content in Markdown format
 */
export type ReadmeGetResponse = string;

export interface ReadmeGetParams {
  /**
   * The name of the user or organization that owns the model.
   */
  model_owner: string;

  /**
   * The name of the model.
   */
  model_name: string;
}

export declare namespace Readme {
  export { type ReadmeGetResponse as ReadmeGetResponse, type ReadmeGetParams as ReadmeGetParams };
}
