// replicate-stainless/resources/files.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Files = void 0;
const resource_1 = require("../core/resource.js");
const pagination_1 = require("../core/pagination.js");
const headers_1 = require("../internal/headers.js");
const uploads_1 = require("../internal/uploads.js");
const path_1 = require("../internal/utils/path.js");
class Files extends resource_1.APIResource {
    /**
     * Create a file by uploading its content and optional metadata.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X POST https://api.replicate.com/v1/files \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   -H 'Content-Type: multipart/form-data' \
     *   -F 'content=@/path/to/archive.zip;type=application/zip;filename=example.zip' \
     *   -F 'metadata={"customer_reference_id": 123};type=application/json'
     * ```
     *
     * The request must include:
     *
     * - `content`: The file content (required)
     * - `type`: The content / MIME type for the file (defaults to
     *   `application/octet-stream`)
     * - `filename`: The filename (required, ≤ 255 bytes, valid UTF-8)
     * - `metadata`: User-provided metadata associated with the file (defaults to `{}`,
     *   must be valid JSON)
     */
    create(body, options) {
        return this._client.post('/files', (0, uploads_1.multipartFormRequestOptions)({ body, ...options }, this._client));
    }
    /**
     * Get a paginated list of all files created by the user or organization associated
     * with the provided API token.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files
     * ```
     *
     * The response will be a paginated JSON array of file objects, sorted with the
     * most recent file first.
     */
    list(options) {
        return this._client.getAPIList('/files', (pagination_1.CursorURLPage), options);
    }
    /**
     * Delete a file. Once a file has been deleted, subsequent requests to the file
     * resource return 404 Not found.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X DELETE \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o
     * ```
     */
    delete(params, options) {
        const { file_id } = params;
        return this._client.delete((0, path_1.path) `/files/${file_id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Download a file by providing the file owner, access expiry, and a valid
     * signature.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X GET "https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o/download?expiry=1708515345&owner=mattt&signature=zuoghqlrcnw8YHywkpaXQlHsVhWen%2FDZ4aal76dLiOo%3D"
     * ```
     */
    download(params, options) {
        const { file_id, ...query } = params;
        return this._client.get((0, path_1.path) `/files/${file_id}/download`, {
            query,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: 'application/octet-stream' }, options?.headers]),
            __binaryResponse: true,
        });
    }
    /**
     * Get the details of a file.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o
     * ```
     */
    get(params, options) {
        const { file_id } = params;
        return this._client.get((0, path_1.path) `/files/${file_id}`, options);
    }
}
exports.Files = Files;
