// replicate-stainless/resources/files.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../core/resource.mjs";
import { CursorURLPage } from "../core/pagination.mjs";
import { buildHeaders } from "../internal/headers.mjs";
import { multipartFormRequestOptions } from "../internal/uploads.mjs";
import { path } from "../internal/utils/path.mjs";
export class Files extends APIResource {
    /**
     * Create a file by uploading its content and optional metadata.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X POST https://api.replicate.com/v1/files \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   -H 'Content-Type: multipart/form-data' \
     *   -F 'content=@/path/to/archive.zip;type=application/zip;filename=example.zip' \
     *   -F 'metadata={"customer_reference_id": 123};type=application/json'
     * ```
     *
     * The request must include:
     *
     * - `content`: The file content (required)
     * - `type`: The content / MIME type for the file (defaults to
     *   `application/octet-stream`)
     * - `filename`: The filename (required, ≤ 255 bytes, valid UTF-8)
     * - `metadata`: User-provided metadata associated with the file (defaults to `{}`,
     *   must be valid JSON)
     */
    create(body, options) {
        return this._client.post('/files', multipartFormRequestOptions({ body, ...options }, this._client));
    }
    /**
     * Get a paginated list of all files created by the user or organization associated
     * with the provided API token.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files
     * ```
     *
     * The response will be a paginated JSON array of file objects, sorted with the
     * most recent file first.
     */
    list(options) {
        return this._client.getAPIList('/files', (CursorURLPage), options);
    }
    /**
     * Delete a file. Once a file has been deleted, subsequent requests to the file
     * resource return 404 Not found.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X DELETE \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o
     * ```
     */
    delete(params, options) {
        const { file_id } = params;
        return this._client.delete(path `/files/${file_id}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Download a file by providing the file owner, access expiry, and a valid
     * signature.
     *
     * Example cURL request:
     *
     * ```console
     * curl -X GET "https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o/download?expiry=1708515345&owner=mattt&signature=zuoghqlrcnw8YHywkpaXQlHsVhWen%2FDZ4aal76dLiOo%3D"
     * ```
     */
    download(params, options) {
        const { file_id, ...query } = params;
        return this._client.get(path `/files/${file_id}/download`, {
            query,
            ...options,
            headers: buildHeaders([{ Accept: 'application/octet-stream' }, options?.headers]),
            __binaryResponse: true,
        });
    }
    /**
     * Get the details of a file.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Token $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/files/cneqzikepnug6xezperrr4z55o
     * ```
     */
    get(params, options) {
        const { file_id } = params;
        return this._client.get(path `/files/${file_id}`, options);
    }
}
