// replicate-stainless/resources/index.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Webhooks = exports.Trainings = exports.Predictions = exports.Models = exports.Hardware = exports.Files = exports.Deployments = exports.Collections = exports.Account = void 0;
var account_1 = require("./account.js");
Object.defineProperty(exports, "Account", { enumerable: true, get: function () { return account_1.Account; } });
var collections_1 = require("./collections.js");
Object.defineProperty(exports, "Collections", { enumerable: true, get: function () { return collections_1.Collections; } });
var deployments_1 = require("./deployments/deployments.js");
Object.defineProperty(exports, "Deployments", { enumerable: true, get: function () { return deployments_1.Deployments; } });
var files_1 = require("./files.js");
Object.defineProperty(exports, "Files", { enumerable: true, get: function () { return files_1.Files; } });
var hardware_1 = require("./hardware.js");
Object.defineProperty(exports, "Hardware", { enumerable: true, get: function () { return hardware_1.Hardware; } });
var models_1 = require("./models/models.js");
Object.defineProperty(exports, "Models", { enumerable: true, get: function () { return models_1.Models; } });
var predictions_1 = require("./predictions.js");
Object.defineProperty(exports, "Predictions", { enumerable: true, get: function () { return predictions_1.Predictions; } });
var trainings_1 = require("./trainings.js");
Object.defineProperty(exports, "Trainings", { enumerable: true, get: function () { return trainings_1.Trainings; } });
var webhooks_1 = require("./webhooks/webhooks.js");
Object.defineProperty(exports, "Webhooks", { enumerable: true, get: function () { return webhooks_1.Webhooks; } });
