// replicate-stainless/src/resources/account.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';

export class Account extends APIResource {
  /**
   * Returns information about the user or organization associated with the provided
   * API token.
   *
   * Example cURL request:
   *
   * ```console
   * curl -s \
   *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
   *   https://api.replicate.com/v1/account
   * ```
   *
   * The response will be a JSON object describing the account:
   *
   * ```json
   * {
   *   "type": "organization",
   *   "username": "acme",
   *   "name": "Acme Corp, Inc.",
   *   "github_url": "https://github.com/acme"
   * }
   * ```
   */
  get(options?: RequestOptions): APIPromise<AccountGetResponse> {
    return this._client.get('/account', options);
  }
}

export interface AccountGetResponse {
  /**
   * The account type. Can be a user or an organization.
   */
  type: 'organization' | 'user';

  /**
   * The username of the account.
   */
  username: string;

  /**
   * The avatar URL for the account.
   */
  avatar_url?: string;

  /**
   * The GitHub URL of the account.
   */
  github_url?: string;

  /**
   * The name of the account.
   */
  name?: string;
}

export declare namespace Account {
  export { type AccountGetResponse as AccountGetResponse };
}
