// replicate-stainless/src/resources/hardware.ts
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.

import { APIResource } from '../core/resource';
import { APIPromise } from '../core/api-promise';
import { RequestOptions } from '../internal/request-options';

export class Hardware extends APIResource {
  /**
   * Example cURL request:
   *
   * ```console
   * curl -s \
   *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
   *   https://api.replicate.com/v1/hardware
   * ```
   *
   * The response will be a JSON array of hardware objects:
   *
   * ```json
   * [
   *   { "name": "CPU", "sku": "cpu" },
   *   { "name": "Nvidia T4 GPU", "sku": "gpu-t4" },
   *   { "name": "Nvidia A40 GPU", "sku": "gpu-a40-small" },
   *   { "name": "Nvidia A40 (Large) GPU", "sku": "gpu-a40-large" }
   * ]
   * ```
   */
  list(options?: RequestOptions): APIPromise<HardwareListResponse> {
    return this._client.get('/hardware', options);
  }
}

export type HardwareListResponse = Array<HardwareListResponse.HardwareListResponseItem>;

export namespace HardwareListResponse {
  export interface HardwareListResponseItem {
    /**
     * The name of the hardware.
     */
    name?: string;

    /**
     * The SKU of the hardware.
     */
    sku?: string;
  }
}

export declare namespace Hardware {
  export { type HardwareListResponse as HardwareListResponse };
}
