// replicate-stainless/resources/models/models.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Models = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const ExamplesAPI = tslib_1.__importStar(require("./examples.js"));
const examples_1 = require("./examples.js");
const PredictionsAPI = tslib_1.__importStar(require("./predictions.js"));
const predictions_1 = require("./predictions.js");
const ReadmeAPI = tslib_1.__importStar(require("./readme.js"));
const readme_1 = require("./readme.js");
const VersionsAPI = tslib_1.__importStar(require("./versions.js"));
const versions_1 = require("./versions.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Models extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.examples = new ExamplesAPI.Examples(this._client);
        this.predictions = new PredictionsAPI.Predictions(this._client);
        this.readme = new ReadmeAPI.Readme(this._client);
        this.versions = new VersionsAPI.Versions(this._client);
    }
    /**
     * Create a model.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s -X POST \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H 'Content-Type: application/json' \
     *   -d '{"owner": "alice", "name": "hot-dog-detector", "description": "Detect hot dogs in images", "visibility": "public", "hardware": "cpu"}' \
     *   https://api.replicate.com/v1/models
     * ```
     *
     * The response will be a model object in the following format:
     *
     * ```json
     * {
     *   "url": "https://replicate.com/alice/hot-dog-detector",
     *   "owner": "alice",
     *   "name": "hot-dog-detector",
     *   "description": "Detect hot dogs in images",
     *   "visibility": "public",
     *   "github_url": null,
     *   "paper_url": null,
     *   "license_url": null,
     *   "run_count": 0,
     *   "cover_image_url": null,
     *   "default_example": null,
     *   "latest_version": null
     * }
     * ```
     *
     * Note that there is a limit of 1,000 models per account. For most purposes, we
     * recommend using a single model and pushing new
     * [versions](https://replicate.com/docs/how-does-replicate-work#versions) of the
     * model as you make changes to it.
     *
     * @example
     * ```ts
     * await replicate.models.create({
     *   hardware: 'cpu',
     *   name: 'hot-dog-detector',
     *   owner: 'alice',
     *   visibility: 'public',
     * });
     * ```
     */
    create(body, options) {
        return this._client.post('/models', {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Get a paginated list of public models.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/models
     * ```
     *
     * The response will be a pagination object containing a list of model objects.
     *
     * See the [`models.get`](#models.get) docs for more details about the model
     * object.
     *
     * @example
     * ```ts
     * // Automatically fetches more pages as needed.
     * for await (const modelListResponse of replicate.models.list()) {
     *   // ...
     * }
     * ```
     */
    list(options) {
        return this._client.getAPIList('/models', (pagination_1.CursorURLPage), options);
    }
    /**
     * Delete a model
     *
     * Model deletion has some restrictions:
     *
     * - You can only delete models you own.
     * - You can only delete private models.
     * - You can only delete models that have no versions associated with them.
     *   Currently you'll need to
     *   [delete the model's versions](#models.versions.delete) before you can delete
     *   the model itself.
     *
     * Example cURL request:
     *
     * ```command
     * curl -s -X DELETE \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/models/replicate/hello-world
     * ```
     *
     * The response will be an empty 204, indicating the model has been deleted.
     *
     * @example
     * ```ts
     * await replicate.models.delete({
     *   model_owner: 'model_owner',
     *   model_name: 'model_name',
     * });
     * ```
     */
    delete(params, options) {
        const { model_owner, model_name } = params;
        return this._client.delete((0, path_1.path) `/models/${model_owner}/${model_name}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/models/replicate/hello-world
     * ```
     *
     * The response will be a model object in the following format:
     *
     * ```json
     * {
     *   "url": "https://replicate.com/replicate/hello-world",
     *   "owner": "replicate",
     *   "name": "hello-world",
     *   "description": "A tiny model that says hello",
     *   "visibility": "public",
     *   "github_url": "https://github.com/replicate/cog-examples",
     *   "paper_url": null,
     *   "license_url": null,
     *   "run_count": 5681081,
     *   "cover_image_url": "...",
     *   "default_example": {...},
     *   "latest_version": {...},
     * }
     * ```
     *
     * The model object includes the
     * [input and output schema](https://replicate.com/docs/reference/openapi#model-schemas)
     * for the latest version of the model.
     *
     * Here's an example showing how to fetch the model with cURL and display its input
     * schema with [jq](https://stedolan.github.io/jq/):
     *
     * ```console
     * curl -s \
     *     -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *     https://api.replicate.com/v1/models/replicate/hello-world \
     *     | jq ".latest_version.openapi_schema.components.schemas.Input"
     * ```
     *
     * This will return the following JSON object:
     *
     * ```json
     * {
     *   "type": "object",
     *   "title": "Input",
     *   "required": ["text"],
     *   "properties": {
     *     "text": {
     *       "type": "string",
     *       "title": "Text",
     *       "x-order": 0,
     *       "description": "Text to prefix with 'hello '"
     *     }
     *   }
     * }
     * ```
     *
     * The `cover_image_url` string is an HTTPS URL for an image file. This can be:
     *
     * - An image uploaded by the model author.
     * - The output file of the example prediction, if the model author has not set a
     *   cover image.
     * - The input file of the example prediction, if the model author has not set a
     *   cover image and the example prediction has no output file.
     * - A generic fallback image.
     *
     * The `default_example` object is a [prediction](#predictions.get) created with
     * this model.
     *
     * The `latest_version` object is the model's most recently pushed
     * [version](#models.versions.get).
     *
     * @example
     * ```ts
     * await replicate.models.get({
     *   model_owner: 'model_owner',
     *   model_name: 'model_name',
     * });
     * ```
     */
    get(params, options) {
        const { model_owner, model_name } = params;
        return this._client.get((0, path_1.path) `/models/${model_owner}/${model_name}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Get a list of public models matching a search query.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s -X QUERY \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H "Content-Type: text/plain" \
     *   -d "hello" \
     *   https://api.replicate.com/v1/models
     * ```
     *
     * The response will be a paginated JSON object containing an array of model
     * objects.
     *
     * See the [`models.get`](#models.get) docs for more details about the model
     * object.
     *
     * @example
     * ```ts
     * await replicate.models.search({ body: 'body' });
     * ```
     */
    search(params, options) {
        const { body } = params;
        return this._client.query('/models', {
            body: body,
            ...options,
            headers: (0, headers_1.buildHeaders)([{ 'Content-Type': 'text/plain', Accept: '*/*' }, options?.headers]),
        });
    }
}
exports.Models = Models;
Models.Examples = examples_1.Examples;
Models.Predictions = predictions_1.Predictions;
Models.Readme = readme_1.Readme;
Models.Versions = versions_1.Versions;
