// replicate-stainless/resources/deployments/deployments.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Deployments = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const PredictionsAPI = tslib_1.__importStar(require("./predictions.js"));
const predictions_1 = require("./predictions.js");
const pagination_1 = require("../../core/pagination.js");
const headers_1 = require("../../internal/headers.js");
const path_1 = require("../../internal/utils/path.js");
class Deployments extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.predictions = new PredictionsAPI.Predictions(this._client);
    }
    /**
     * Create a new deployment:
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -X POST \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H "Content-Type: application/json" \
     *   -d '{
     *         "name": "my-app-image-generator",
     *         "model": "stability-ai/sdxl",
     *         "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *         "hardware": "gpu-t4",
     *         "min_instances": 0,
     *         "max_instances": 3
     *       }' \
     *   https://api.replicate.com/v1/deployments
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 1,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 1,
     *       "max_instances": 5
     *     }
     *   }
     * }
     * ```
     */
    create(body, options) {
        return this._client.post('/deployments', { body, ...options });
    }
    /**
     * Update properties of an existing deployment, including hardware, min/max
     * instances, and the deployment's underlying model
     * [version](https://replicate.com/docs/how-does-replicate-work#versions).
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -X PATCH \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H "Content-Type: application/json" \
     *   -d '{"min_instances": 3, "max_instances": 10}' \
     *   https://api.replicate.com/v1/deployments/acme/my-app-image-generator
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 2,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 3,
     *       "max_instances": 10
     *     }
     *   }
     * }
     * ```
     *
     * Updating any deployment properties will increment the `number` field of the
     * `current_release`.
     */
    update(params, options) {
        const { deployment_owner, deployment_name, ...body } = params;
        return this._client.patch((0, path_1.path) `/deployments/${deployment_owner}/${deployment_name}`, {
            body,
            ...options,
        });
    }
    /**
     * Get a list of deployments associated with the current account, including the
     * latest release configuration for each deployment.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments
     * ```
     *
     * The response will be a paginated JSON array of deployment objects, sorted with
     * the most recent deployment first:
     *
     * ```json
     * {
     *   "next": "http://api.replicate.com/v1/deployments?cursor=cD0yMDIzLTA2LTA2KzIzJTNBNDAlM0EwOC45NjMwMDAlMkIwMCUzQTAw",
     *   "previous": null,
     *   "results": [
     *     {
     *       "owner": "replicate",
     *       "name": "my-app-image-generator",
     *       "current_release": {
     *         "number": 1,
     *         "model": "stability-ai/sdxl",
     *         "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *         "created_at": "2024-02-15T16:32:57.018467Z",
     *         "created_by": {
     *           "type": "organization",
     *           "username": "acme",
     *           "name": "Acme Corp, Inc.",
     *           "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *           "github_url": "https://github.com/acme"
     *         },
     *         "configuration": {
     *           "hardware": "gpu-t4",
     *           "min_instances": 1,
     *           "max_instances": 5
     *         }
     *       }
     *     }
     *   ]
     * }
     * ```
     */
    list(options) {
        return this._client.getAPIList('/deployments', (pagination_1.CursorURLPage), options);
    }
    /**
     * Delete a deployment
     *
     * Deployment deletion has some restrictions:
     *
     * - You can only delete deployments that have been offline and unused for at least
     *   15 minutes.
     *
     * Example cURL request:
     *
     * ```command
     * curl -s -X DELETE \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments/acme/my-app-image-generator
     * ```
     *
     * The response will be an empty 204, indicating the deployment has been deleted.
     */
    delete(params, options) {
        const { deployment_owner, deployment_name } = params;
        return this._client.delete((0, path_1.path) `/deployments/${deployment_owner}/${deployment_name}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Get information about a deployment by name including the current release.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments/replicate/my-app-image-generator
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 1,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 1,
     *       "max_instances": 5
     *     }
     *   }
     * }
     * ```
     */
    get(params, options) {
        const { deployment_owner, deployment_name } = params;
        return this._client.get((0, path_1.path) `/deployments/${deployment_owner}/${deployment_name}`, options);
    }
}
exports.Deployments = Deployments;
Deployments.Predictions = predictions_1.Predictions;
