// replicate-stainless/resources/deployments/deployments.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import * as PredictionsAPI from "./predictions.mjs";
import { Predictions } from "./predictions.mjs";
import { CursorURLPage } from "../../core/pagination.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Deployments extends APIResource {
    constructor() {
        super(...arguments);
        this.predictions = new PredictionsAPI.Predictions(this._client);
    }
    /**
     * Create a new deployment:
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -X POST \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H "Content-Type: application/json" \
     *   -d '{
     *         "name": "my-app-image-generator",
     *         "model": "stability-ai/sdxl",
     *         "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *         "hardware": "gpu-t4",
     *         "min_instances": 0,
     *         "max_instances": 3
     *       }' \
     *   https://api.replicate.com/v1/deployments
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 1,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 1,
     *       "max_instances": 5
     *     }
     *   }
     * }
     * ```
     */
    create(body, options) {
        return this._client.post('/deployments', { body, ...options });
    }
    /**
     * Update properties of an existing deployment, including hardware, min/max
     * instances, and the deployment's underlying model
     * [version](https://replicate.com/docs/how-does-replicate-work#versions).
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -X PATCH \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   -H "Content-Type: application/json" \
     *   -d '{"min_instances": 3, "max_instances": 10}' \
     *   https://api.replicate.com/v1/deployments/acme/my-app-image-generator
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 2,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 3,
     *       "max_instances": 10
     *     }
     *   }
     * }
     * ```
     *
     * Updating any deployment properties will increment the `number` field of the
     * `current_release`.
     */
    update(params, options) {
        const { deployment_owner, deployment_name, ...body } = params;
        return this._client.patch(path `/deployments/${deployment_owner}/${deployment_name}`, {
            body,
            ...options,
        });
    }
    /**
     * Get a list of deployments associated with the current account, including the
     * latest release configuration for each deployment.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments
     * ```
     *
     * The response will be a paginated JSON array of deployment objects, sorted with
     * the most recent deployment first:
     *
     * ```json
     * {
     *   "next": "http://api.replicate.com/v1/deployments?cursor=cD0yMDIzLTA2LTA2KzIzJTNBNDAlM0EwOC45NjMwMDAlMkIwMCUzQTAw",
     *   "previous": null,
     *   "results": [
     *     {
     *       "owner": "replicate",
     *       "name": "my-app-image-generator",
     *       "current_release": {
     *         "number": 1,
     *         "model": "stability-ai/sdxl",
     *         "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *         "created_at": "2024-02-15T16:32:57.018467Z",
     *         "created_by": {
     *           "type": "organization",
     *           "username": "acme",
     *           "name": "Acme Corp, Inc.",
     *           "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *           "github_url": "https://github.com/acme"
     *         },
     *         "configuration": {
     *           "hardware": "gpu-t4",
     *           "min_instances": 1,
     *           "max_instances": 5
     *         }
     *       }
     *     }
     *   ]
     * }
     * ```
     */
    list(options) {
        return this._client.getAPIList('/deployments', (CursorURLPage), options);
    }
    /**
     * Delete a deployment
     *
     * Deployment deletion has some restrictions:
     *
     * - You can only delete deployments that have been offline and unused for at least
     *   15 minutes.
     *
     * Example cURL request:
     *
     * ```command
     * curl -s -X DELETE \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments/acme/my-app-image-generator
     * ```
     *
     * The response will be an empty 204, indicating the deployment has been deleted.
     */
    delete(params, options) {
        const { deployment_owner, deployment_name } = params;
        return this._client.delete(path `/deployments/${deployment_owner}/${deployment_name}`, {
            ...options,
            headers: buildHeaders([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Get information about a deployment by name including the current release.
     *
     * Example cURL request:
     *
     * ```console
     * curl -s \
     *   -H "Authorization: Bearer $REPLICATE_API_TOKEN" \
     *   https://api.replicate.com/v1/deployments/replicate/my-app-image-generator
     * ```
     *
     * The response will be a JSON object describing the deployment:
     *
     * ```json
     * {
     *   "owner": "acme",
     *   "name": "my-app-image-generator",
     *   "current_release": {
     *     "number": 1,
     *     "model": "stability-ai/sdxl",
     *     "version": "da77bc59ee60423279fd632efb4795ab731d9e3ca9705ef3341091fb989b7eaf",
     *     "created_at": "2024-02-15T16:32:57.018467Z",
     *     "created_by": {
     *       "type": "organization",
     *       "username": "acme",
     *       "name": "Acme Corp, Inc.",
     *       "avatar_url": "https://cdn.replicate.com/avatars/acme.png",
     *       "github_url": "https://github.com/acme"
     *     },
     *     "configuration": {
     *       "hardware": "gpu-t4",
     *       "min_instances": 1,
     *       "max_instances": 5
     *     }
     *   }
     * }
     * ```
     */
    get(params, options) {
        const { deployment_owner, deployment_name } = params;
        return this._client.get(path `/deployments/${deployment_owner}/${deployment_name}`, options);
    }
}
Deployments.Predictions = Predictions;
