// replicate-stainless/src/lib/run.ts
import { Replicate } from '../client';
import { Prediction, PredictionCreateParams } from '../resources';
import { ModelVersionIdentifier } from './identifier';
import { createFileOutput, transform } from './util';
import { toBase64 } from '../internal/utils/base64';
import { checkFileSupport } from '../internal/uploads';
import { APIError } from '../core/error';

export type RunOptions = {
  wait?: { mode: 'block' | 'poll'; timeout?: number; interval?: number } | false;
  useFileOutput?: boolean;
  /**
   * Strategy for handling file inputs (File, Blob, Buffer objects):
   * - 'default': Upload files to get URLs, fall back to base64 on server errors
   * - 'upload': Always upload files to get URLs
   * - 'data-uri': Always encode files as base64 data URIs
   */
  fileEncodingStrategy?: 'data-uri' | 'upload' | 'default';
  signal?: AbortSignal;
} & PredictionCreateParams;

/**
 * Transforms file inputs according to the specified strategy.
 */
async function transformFileInputs(
  client: Replicate,
  inputs: unknown,
  strategy: 'data-uri' | 'upload' | 'default' = 'default',
): Promise<unknown> {
  switch (strategy) {
    case 'data-uri':
      return await transformFileInputsToBase64EncodedDataURIs(inputs);
    case 'upload':
      return await transformFileInputsToReplicateFileURLs(client, inputs);
    case 'default':
      try {
        return await transformFileInputsToReplicateFileURLs(client, inputs);
      } catch (error) {
        // If upload fails with client error (4xx), re-throw
        if (error instanceof APIError && error.status >= 400 && error.status < 500) {
          throw error;
        }
        // Otherwise fall back to base64 encoding for server errors or other failures
        return await transformFileInputsToBase64EncodedDataURIs(inputs);
      }
    default:
      throw new Error(`Unexpected file upload strategy: ${strategy}`);
  }
}

/**
 * Uploads File/Blob/Buffer objects to Replicate and replaces them with URLs.
 */
async function transformFileInputsToReplicateFileURLs(client: Replicate, inputs: unknown): Promise<unknown> {
  return await transform(inputs, async (value: unknown) => {
    if (value instanceof Blob || Buffer.isBuffer(value)) {
      if (Buffer.isBuffer(value)) {
        checkFileSupport();
        const file = new File([value], 'file', { type: 'application/octet-stream' });
        const fileResponse = await client.files.create({ content: file });
        return fileResponse.urls.get;
      }
      const fileResponse = await client.files.create({ content: value });
      return fileResponse.urls.get;
    }

    return value;
  });
}

const MAX_DATA_URI_SIZE = 10_000_000;

/**
 * Check if a value is a TypedArray
 */
function isTypedArray(value: unknown): value is ArrayBufferView {
  return (
    value instanceof Int8Array ||
    value instanceof Int16Array ||
    value instanceof Int32Array ||
    value instanceof Uint8Array ||
    value instanceof Uint8ClampedArray ||
    value instanceof Uint16Array ||
    value instanceof Uint32Array ||
    value instanceof Float32Array ||
    value instanceof Float64Array
  );
}

/**
 * Converts File/Blob/Buffer objects to base64-encoded data URIs.
 */
async function transformFileInputsToBase64EncodedDataURIs(inputs: unknown): Promise<unknown> {
  let totalBytes = 0;

  return await transform(inputs, async (value: unknown) => {
    let buffer: ArrayBuffer | ArrayBufferView;
    let mime: string | undefined;

    if (value instanceof Blob) {
      buffer = await value.arrayBuffer();
      mime = value.type;
    } else if (isTypedArray(value)) {
      buffer = value;
    } else {
      return value;
    }

    totalBytes += buffer.byteLength;
    if (totalBytes > MAX_DATA_URI_SIZE) {
      throw new Error(
        `Combined filesize of prediction ${totalBytes} bytes exceeds 10mb limit for inline encoding, please provide URLs instead`,
      );
    }

    // Convert to Uint8Array for base64 encoding
    let uint8Array: Uint8Array;
    if (buffer instanceof ArrayBuffer) {
      uint8Array = new Uint8Array(buffer);
    } else if (buffer instanceof Uint8Array) {
      uint8Array = buffer;
    } else {
      // For other typed arrays, create a Uint8Array view
      uint8Array = new Uint8Array(buffer.buffer, buffer.byteOffset, buffer.byteLength);
    }

    const base64 = toBase64(uint8Array);
    mime = mime || 'application/octet-stream';

    return `data:${mime};base64,${base64}`;
  });
}

export async function run(
  client: Replicate,
  fetch: typeof globalThis.fetch,
  ref: string,
  options: RunOptions,
  progress?: (prediction: Prediction) => void,
) {
  const identifier = ModelVersionIdentifier.parse(ref);
  const { wait, useFileOutput, fileEncodingStrategy, signal, ...params } = options;

  if (wait === false && params?.Prefer) {
    throw new Error('Cannot mix and match prefer and wait');
  }

  if (typeof wait === 'object' && wait.mode === 'block') {
    params.Prefer = wait.timeout ? `wait=${wait.timeout}` : 'wait';
  }

  if (params.input) {
    params.input = await transformFileInputs(client, params.input, fileEncodingStrategy);
  }

  let prediction: Prediction;
  if (identifier.version) {
    prediction = await client.predictions.create({
      ...params,
      version: identifier.version,
    });
  } else if (identifier.owner && identifier.name) {
    const modelPredictionParams = {
      model_owner: identifier.owner,
      model_name: identifier.name,
      ...params,
    };
    prediction = await client.models.predictions.create(modelPredictionParams);
  } else {
    throw new Error('Invalid model version identifier');
  }

  // Call progress callback with the initial prediction object
  if (progress) {
    progress(prediction);
  }

  const isDone = wait && wait.mode === 'block' && prediction.status !== 'starting';
  if (!isDone) {
    prediction = await client.wait(
      prediction,
      { interval: wait && wait.mode === 'poll' ? wait.interval : undefined },
      async (updatedPrediction: Prediction) => {
        // Call progress callback with the updated prediction object
        if (progress) {
          progress(updatedPrediction);
        }

        // We handle the cancel later in the function.
        if (signal && signal.aborted) {
          return true; // stop polling
        }

        return false; // continue polling
      },
    );
  }

  if (signal && signal.aborted) {
    prediction = await client.predictions.cancel({
      prediction_id: prediction.id,
    });
  }

  // Call progress callback with the completed prediction object
  if (progress) {
    progress(prediction);
  }

  if (prediction.status === 'failed') {
    throw new Error(`Prediction failed: ${prediction.error}`);
  }
  return await transform(prediction.output as object, (value: any) => {
    if (typeof value === 'string' && (value.startsWith('https:') || value.startsWith('data:'))) {
      return useFileOutput ? createFileOutput({ url: value, fetch }) : value;
    }
    return value;
  });
}
