// replicate-stainless/src/lib/util.ts
import { APIError } from '../error';

// Walk a JavaScript object and transform the leaf values.
export async function transform(value: any, mapper: (value: any) => any): Promise<any> {
  if (Array.isArray(value)) {
    const copy = [];
    for (const val of value) {
      const transformed = await transform(val, mapper);
      copy.push(transformed);
    }
    return copy;
  }

  if (isPlainObject(value)) {
    const copy: Record<string, any> = {};
    for (const key of Object.keys(value)) {
      copy[key] = await transform(value[key], mapper);
    }
    return copy;
  }

  return await mapper(value);
}

export interface FileOutput extends ReadableStream {
  blob(): Promise<Blob>;
  url(): URL;
  toString(): string;
}

export function createFileOutput({ url, fetch }: { url: string; fetch: (url: string) => Promise<Response> }) {
  let type = 'application/octet-stream';

  class FileOutput<T extends Blob | ArrayBuffer = Blob> extends ReadableStream<T> {
    async blob() {
      const chunks: T[] = [];
      const iterator = this[Symbol.asyncIterator]();
      let result = await iterator.next();

      while (!result.done) {
        chunks.push(result.value);
        result = await iterator.next();
      }

      return new Blob(chunks, { type });
    }

    url() {
      return new URL(url);
    }

    override toString() {
      return url;
    }
  }

  return new FileOutput({
    async start(controller) {
      const response = await fetch(url);

      if (!response.ok) {
        const text = await response.text();
        controller.error(
          new APIError(
            response.status,
            {},
            `Request to ${url} failed with status ${response.status}: ${text}`,
            response.headers,
          ),
        );
      }

      if (response.headers.get('Content-Type')) {
        type = response.headers.get('Content-Type')!;
      }

      try {
        for await (const chunk of streamAsyncIterator(response.body!)) {
          controller.enqueue(chunk);
        }
        controller.close();
      } catch (err) {
        controller.error(err);
      }
    },
  });
}

// Test for a plain JS object.
// Source: lodash.isPlainObject
function isPlainObject(value: unknown): value is Record<string, unknown> {
  const isObjectLike = typeof value === 'object' && value !== null;
  if (!isObjectLike || String(value) !== '[object Object]') {
    return false;
  }
  const proto = Object.getPrototypeOf(value);
  if (proto === null) {
    return true;
  }
  const Ctor = Object.prototype.hasOwnProperty.call(proto, 'constructor') && proto.constructor;
  return (
    typeof Ctor === 'function' &&
    Ctor instanceof Ctor &&
    Function.prototype.toString.call(Ctor) === Function.prototype.toString.call(Object)
  );
}

/**
 * Helper to make any `ReadableStream` iterable, this is supported
 * by most server runtimes but browsers still haven't implemented
 * it yet.
 * See: https://developer.mozilla.org/en-US/docs/Web/API/ReadableStream#browser_compatibility
 *
 * @template T
 * @param {ReadableStream<T>} stream an instance of a `ReadableStream`
 * @yields {T} a chunk/event from the stream
 */
async function* streamAsyncIterator<T>(stream: ReadableStream<T>) {
  const reader = stream.getReader();
  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) return;
      yield value;
    }
  } finally {
    reader.releaseLock();
  }
}

/**
 * Parse progress from prediction logs.
 *
 * This function supports log statements in the following format,
 * which are generated by https://github.com/tqdm/tqdm and similar libraries:
 *
 * ```
 * 76%|████████████████████████████         | 7568/10000 [00:33<00:10, 229.00it/s]
 * ```
 *
 * @example
 * const progress = parseProgressFromLogs("76%|████████████████████████████         | 7568/10000 [00:33<00:10, 229.00it/s]");
 * console.log(progress);
 * // {
 * //   percentage: 0.76,
 * //   current: 7568,
 * //   total: 10000,
 * // }
 *
 * @param {object|string} input - A prediction object or string.
 * @returns {(object|null)} - An object with the percentage, current, and total, or null if no progress can be parsed.
 */
export function parseProgressFromLogs(input: { logs?: string } | string): object | null {
  const logs = typeof input === 'object' && 'logs' in input && input.logs ? input.logs : input;
  if (!logs || typeof logs !== 'string') {
    return null;
  }

  const pattern = /^\s*(\d+)%\s*\|.+?\|\s*(\d+)\/(\d+)/;
  const lines = logs.split('\n').reverse();

  for (const line of lines) {
    const matches = line.match(pattern);

    if (matches && matches.length === 4) {
      return {
        percentage: parseInt(matches[1]!, 10) / 100,
        current: parseInt(matches[2]!, 10),
        total: parseInt(matches[3]!, 10),
      };
    }
  }

  return null;
}
