// replicate-stainless/src/lib/wait.ts
import { sleep } from '../internal/utils';
import { Replicate } from '../client';
import { Prediction } from '../resources';

export async function wait(
  client: Replicate,
  prediction: Prediction,
  options: { interval?: number | undefined },
  stop: (prediction: Prediction) => boolean | Promise<boolean>,
): Promise<Prediction> {
  const { id } = prediction;
  if (!id) {
    throw new Error('Invalid prediction');
  }

  if (
    prediction.status === 'succeeded' ||
    prediction.status === 'failed' ||
    prediction.status === 'canceled'
  ) {
    return prediction;
  }

  const interval = (options && options.interval) || 500;

  let updatedPrediction = await client.predictions.get({
    prediction_id: id,
  });

  while (
    updatedPrediction.status !== 'succeeded' &&
    updatedPrediction.status !== 'failed' &&
    updatedPrediction.status !== 'canceled'
  ) {
    if (stop && (await stop(updatedPrediction)) === true) {
      break;
    }

    await sleep(interval);
    updatedPrediction = await client.predictions.get({
      prediction_id: id,
    });
  }

  if (updatedPrediction.status === 'failed') {
    throw new Error(`Prediction failed: ${updatedPrediction.error}`);
  }

  return updatedPrediction;
}
