// replicate-stainless/core/pagination.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
var _AbstractPage_client;
import { __classPrivateFieldGet, __classPrivateFieldSet } from "../internal/tslib.mjs";
import { ReplicateError } from "./error.mjs";
import { defaultParseResponse } from "../internal/parse.mjs";
import { APIPromise } from "./api-promise.mjs";
export class AbstractPage {
    constructor(client, response, body, options) {
        _AbstractPage_client.set(this, void 0);
        __classPrivateFieldSet(this, _AbstractPage_client, client, "f");
        this.options = options;
        this.response = response;
        this.body = body;
    }
    hasNextPage() {
        const items = this.getPaginatedItems();
        if (!items.length)
            return false;
        return this.nextPageRequestOptions() != null;
    }
    async getNextPage() {
        const nextOptions = this.nextPageRequestOptions();
        if (!nextOptions) {
            throw new ReplicateError('No next page expected; please check `.hasNextPage()` before calling `.getNextPage()`.');
        }
        return await __classPrivateFieldGet(this, _AbstractPage_client, "f").requestAPIList(this.constructor, nextOptions);
    }
    async *iterPages() {
        let page = this;
        yield page;
        while (page.hasNextPage()) {
            page = await page.getNextPage();
            yield page;
        }
    }
    async *[(_AbstractPage_client = new WeakMap(), Symbol.asyncIterator)]() {
        for await (const page of this.iterPages()) {
            for (const item of page.getPaginatedItems()) {
                yield item;
            }
        }
    }
}
/**
 * This subclass of Promise will resolve to an instantiated Page once the request completes.
 *
 * It also implements AsyncIterable to allow auto-paginating iteration on an unawaited list call, eg:
 *
 *    for await (const item of client.items.list()) {
 *      console.log(item)
 *    }
 */
export class PagePromise extends APIPromise {
    constructor(client, request, Page) {
        super(client, request, async (client, props) => new Page(client, props.response, await defaultParseResponse(client, props), props.options));
    }
    /**
     * Allow auto-paginating iteration on an unawaited list call, eg:
     *
     *    for await (const item of client.items.list()) {
     *      console.log(item)
     *    }
     */
    async *[Symbol.asyncIterator]() {
        const page = await this;
        for await (const item of page) {
            yield item;
        }
    }
}
export class CursorURLPageWithCreatedFilters extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.results = body.results || [];
        this.next = body.next || '';
    }
    getPaginatedItems() {
        return this.results ?? [];
    }
    nextPageRequestOptions() {
        const urlString = this.next;
        if (!urlString)
            return null;
        const url = new URL(urlString);
        // ensure all of the query params that we used for the last request are also included
        // in the next request
        const params = [...Object.entries(this.options.query || {}), ...url.searchParams.entries()];
        for (const [key, value] of params) {
            url.searchParams.set(key, value);
        }
        return {
            ...this.options,
            path: url.toString(),
            query: undefined,
        };
    }
}
export class CursorURLPage extends AbstractPage {
    constructor(client, response, body, options) {
        super(client, response, body, options);
        this.results = body.results || [];
        this.next = body.next || '';
    }
    getPaginatedItems() {
        return this.results ?? [];
    }
    nextPageRequestOptions() {
        const urlString = this.next;
        if (!urlString)
            return null;
        const url = new URL(urlString);
        // ensure all of the query params that we used for the last request are also included
        // in the next request
        const params = [...Object.entries(this.options.query || {}), ...url.searchParams.entries()];
        for (const [key, value] of params) {
            url.searchParams.set(key, value);
        }
        return {
            ...this.options,
            path: url.toString(),
            query: undefined,
        };
    }
}
